File: Exercicios_Verde/pilha/pilha_java/Pilha.java
import java.util.Scanner;

class Pilha {
	private int arr[];
	private int n;

	public Pilha(int tamanho) {
		arr = new int[tamanho];
		n = 0;
	} //constroi a pilha com capacidade máxima igual a tamanho
	
	// FI
	public void empilhar(int x) {
		if (n < arr.length) {
			arr[n] = x;
			n++;
		}	
	} //empilha x à pilha

	public int desempilhar() {
		if (n > 0) {
			return arr[--n];
		}		
		return -1;	
	} //desempilha da pilha e retorna o elemento desempilhado

	public void mostrar() {

		if (n > 0) {
			for (int i = n-1; i >= 0; i--) {
				System.out.print("" + arr[i] + " ");
			}
		} else System.out.print("V");

		System.out.println("");
	} //imprime os elementos da pilha partindo do topo

	public boolean pesquisar(int x) { 
		boolean res = false;
		
		for (int i = 0; i < n; i++) {
			if (arr[i] == x) {res = true; i = n;}
		}

		return res;
	} //retorna true caso o elemento x esteja na pilha e false caso contrário

	public static void main(String[] args) {
		Scanner inp = new Scanner(System.in);
		char comando = '0';
		Pilha pilha = new Pilha(1);

		while (inp.hasNext()) {
			comando = inp.next().charAt(0);
			if (comando == 'C') pilha = new Pilha(inp.nextInt());
			else if (comando == 'E') pilha.empilhar(inp.nextInt());
			else if (comando == 'D') System.out.println(pilha.desempilhar());
			else if (comando == 'M') pilha.mostrar();
			else {
				if (pilha.pesquisar(inp.nextInt())) {
					System.out.println("S");
				} else System.out.println("N");
			}
		}
	}
}

File: Exercicios_Verde/pilha/pilha_c/pilha.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct {
	int * arr;
	int n;
	int cap;
} Pilha;

Pilha * newPilha(int cap) {
	Pilha * p = NULL;

	if (cap > 0) {
		p = malloc(sizeof(Pilha));
		p->arr = malloc(cap*sizeof(int));
		p->n = 0;
		p->cap = cap;
	}
	return p;	
}

void delPilha(Pilha* p) { 
	free(p->arr);
	free(p);
} 

void empilhar(Pilha* p, int x) { 
	if (p->n < p->cap) {
		p->arr[p->n] = x;
		p->n++;
	}
}

int desempilhar(Pilha* p) { 
	int res = -1;
	
	if (p->n >= 0) {
		return p->arr[--p->n];
	}	

	return res;
} 


void mostrar(Pilha* p) { 
	for (int i = p->n-1; i >=0; i--) {
		printf("%d ", p->arr[i]);
	}

	printf("\n");
} 

bool pesquisar(Pilha* p, int x) {
	bool res = false;
	for (int i = 0; i < p->n; i++) {
		if (p->arr[i] == x) {res = true; i = p->n;}
	} 
	return res;
}

void main () {
	Pilha * p = NULL;
	char comando = '0';
	int n = 0;

	while (scanf("%c", &comando) == 1) {
		if (comando == 'C') {
			scanf("%d", &n);
			p = newPilha(n);
		} else if (comando == 'E') {
			scanf("%d", &n);
			empilhar(p, n);
		} else if (comando == 'P') {
			scanf("%d", &n);
			if (pesquisar(p, n)) printf("%c\n", 'S');
			else printf("%c\n", 'N'); 
		} else if (comando == 'D') printf("%d\n", desempilhar(p));
		else if (comando == 'M') mostrar(p);
		
	}

}

